File: app/src/main/java/com/simonc312/androidapiexercise/UpcomingGuidePresenter.java
package com.simonc312.androidapiexercise;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import com.simonc312.androidapiexercise.api.models.Guide;

import java.util.List;

/**
 * Presents {@link Guide}
 * data in displayable form through view contract.
 */

public class UpcomingGuidePresenter implements GuideInteractor.InteractorOutput {

    private final GuideInteractor guideInteractor;
    private final View view;

    public UpcomingGuidePresenter(@NonNull final GuideInteractor guideInteractor,
                                  @NonNull final View view) {
        this.guideInteractor = guideInteractor;
        this.view = view;
    }

    /**
     * Make request to get upcoming guide date
     */
    public void get() {
        this.guideInteractor.get();
    }

    /**
     * Cancel existing request.
     */
    public void cancel() {
        this.guideInteractor.cancel();
    }

    //region InteractorOutput
    @Override
    public void onGuidesAvailable(List<Guide> guides) {
        this.view.update(guides);
    }

    @Override
    public void onGuidesAvailableOffline(List<Guide> guides) {
        this.view.update(guides);
        this.view.showToast(R.string.upcoming_guide_offline_mode);
    }

    @Override
    public void onGuidesUnavailable() {
        this.view.showToast(R.string.upcoming_guide_request_error);
    }
    //endregion

    public interface View {
        void update(@NonNull List<Guide> newGuides);

        void showToast(@StringRes int messageRes);
    }
}

File: app/src/main/java/com/simonc312/androidapiexercise/UpcomingGuidePresenter.kt
package com.simonc312.androidapiexercise

import android.support.annotation.StringRes

import com.simonc312.androidapiexercise.api.models.Guide

/**
 * Presents [Guide]
 * data in displayable form through view contract.
 */

class UpcomingGuidePresenter(private val guideInteractor: GuideInteractor,
                             private val view: UpcomingGuidePresenter.View)
    //implements
    : GuideInteractor.InteractorOutput {

    /**
     * Make request to get upcoming guide date
     */
    fun get() {
        this.guideInteractor.get()
    }

    fun get(query: String) {
        this.guideInteractor.get(query)
    }

    /**
     * Cancel existing request.
     */
    fun cancel() {
        this.guideInteractor.cancel()
    }

    //region InteractorOutput
    override fun onGuidesAvailable(guides: List<Guide>) {
        this.view.update(guides)
    }

    override fun onGuidesAvailableOffline(guides: List<Guide>) {
        this.view.update(guides)
        this.view.showToast(R.string.upcoming_guide_offline_mode)
    }

    override fun onGuidesUnavailable() {
        this.view.showToast(R.string.upcoming_guide_request_error)
    }

    //endregion

    interface View {
        fun update(newGuides: List<Guide>)

        fun showToast(@StringRes messageRes: Int)
    }
}
